File: src/main/java/com/owncloud/android/ui/preview/FileDownloadFragment.java
/*
 *   ownCloud Android client application
 *
 *   @author David A. Velasco
 *   Copyright (C) 2015  ownCloud Inc.
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License version 2,
 *   as published by the Free Software Foundation.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.owncloud.android.ui.preview;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.nextcloud.client.account.User;
import com.owncloud.android.R;
import com.owncloud.android.datamodel.OCFile;
import com.owncloud.android.lib.common.network.OnDatatransferProgressListener;
import com.owncloud.android.lib.common.utils.Log_OC;
import com.owncloud.android.ui.fragment.FileFragment;
import com.owncloud.android.utils.theme.ThemeBarUtils;
import com.owncloud.android.utils.theme.ThemeColorUtils;

import java.lang.ref.WeakReference;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentStatePagerAdapter;


/**
 * This Fragment is used to monitor the progress of a file downloading.
 */
public class FileDownloadFragment extends FileFragment implements OnClickListener {

    public static final String EXTRA_FILE = "FILE";
    public static final String EXTRA_USER = "USER";
    private static final String EXTRA_ERROR = "ERROR";

    private static final String ARG_FILE = "FILE";
    private static final String ARG_IGNORE_FIRST = "IGNORE_FIRST";
    private static final String ARG_USER = "USER" ;

    private View mView;
    private User user;

    public ProgressListener mProgressListener;
    private boolean mListening;

    private static final String TAG = FileDownloadFragment.class.getSimpleName();

    private boolean mIgnoreFirstSavedState;
    private boolean mError;


    /**
     * Public factory method to create a new fragment that shows the progress of a file download.
     *
     * Android strongly recommends keep the empty constructor of fragments as the only public constructor, and
     * use {@link #setArguments(Bundle)} to set the needed arguments.
     *
     * This method hides to client objects the need of doing the construction in two steps.
     *
     * When 'file' is null creates a dummy layout (useful when a file wasn't tapped before).
     *
     * @param file                      An {@link OCFile} to show in the fragment
     * @param user                      Nextcloud user; needed to start downloads
     * @param ignoreFirstSavedState     Flag to work around an unexpected behaviour of {@link FragmentStatePagerAdapter}
     *                                  TODO better solution
     */
    public static Fragment newInstance(OCFile file, User user, boolean ignoreFirstSavedState) {
        FileDownloadFragment frag = new FileDownloadFragment();
        Bundle args = new Bundle();
        args.putParcelable(ARG_FILE, file);
        args.putParcelable(ARG_USER, user);
        args.putBoolean(ARG_IGNORE_FIRST, ignoreFirstSavedState);
        frag.setArguments(args);
        return frag;
    }


    /**
     * Creates an empty details fragment.
     *
     * It's necessary to keep a public constructor without parameters; the system uses it when tries to
     * reinstantiate a fragment automatically.
     */
    public FileDownloadFragment() {
        super();
        mProgressListener = null;
        mListening = false;
        mIgnoreFirstSavedState = false;
        mError = false;
    }


    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        setFile(args.getParcelable(ARG_FILE));
            // TODO better in super, but needs to check ALL the class extending FileFragment; not right now

        mIgnoreFirstSavedState = args.getBoolean(ARG_IGNORE_FIRST);
        user = args.getParcelable(ARG_USER);
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        if (savedInstanceState != null) {
            if (!mIgnoreFirstSavedState) {
                setFile(savedInstanceState.getParcelable(EXTRA_FILE));
                user = savedInstanceState.getParcelable(EXTRA_USER);
                mError = savedInstanceState.getBoolean(EXTRA_ERROR);
            }
            else {
                mIgnoreFirstSavedState = false;
            }
        }

        mView = inflater.inflate(R.layout.file_download_fragment, container, false);

        ProgressBar progressBar = mView.findViewById(R.id.progressBar);
        ThemeBarUtils.colorHorizontalProgressBar(progressBar, ThemeColorUtils.primaryAccentColor(getContext()));
        mProgressListener = new ProgressListener(progressBar);

        (mView.findViewById(R.id.cancelBtn)).setOnClickListener(this);

        (mView.findViewById(R.id.fileDownloadLL)).setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                ((PreviewImageActivity) getActivity()).toggleFullScreen();
            }
        });

        if (mError) {
            setButtonsForRemote();
        }
        else {
            setButtonsForTransferring();
        }

        return mView;
    }


    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(FileDownloadFragment.EXTRA_FILE, getFile());
        outState.putParcelable(FileDownloadFragment.EXTRA_USER, user);
        outState.putBoolean(FileDownloadFragment.EXTRA_ERROR, mError);
    }

    @Override
    public void onStart() {
        super.onStart();
        listenForTransferProgress();
    }

    @Override
    public void onResume() {
        super.onResume();
    }


    @Override
    public void onPause() {
        super.onPause();
    }


    @Override
    public void onStop() {
        leaveTransferProgress();
        super.onStop();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }


    @Override
    public View getView() {
        if (!mListening) {
            listenForTransferProgress();
        }
        return super.getView() == null ? mView : super.getView();
    }


    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.cancelBtn) {
            containerActivity.getFileOperationsHelper().cancelTransference(getFile());
            getActivity().finish();
        } else {
            Log_OC.e(TAG, "Incorrect view clicked!");
        }
    }


    /**
     * Enables or disables buttons for a file being downloaded
     */
    private void setButtonsForTransferring() {
        getView().findViewById(R.id.cancelBtn).setVisibility(View.VISIBLE);

        // show the progress bar for the transfer
        getView().findViewById(R.id.progressBar).setVisibility(View.VISIBLE);
        TextView progressText = getView().findViewById(R.id.progressText);
        progressText.setText(R.string.downloader_download_in_progress_ticker);
        progressText.setVisibility(View.VISIBLE);

        // hides the error icon
        getView().findViewById(R.id.errorText).setVisibility(View.GONE);
        getView().findViewById(R.id.error_image).setVisibility(View.GONE);
    }

    /**
     * Enables or disables buttons for a file not locally available
     *
     * Currently, this is only used when a download was failed
     */
    private void setButtonsForRemote() {
        getView().findViewById(R.id.cancelBtn).setVisibility(View.GONE);

        // hides the progress bar and message
        getView().findViewById(R.id.progressBar).setVisibility(View.GONE);
        getView().findViewById(R.id.progressText).setVisibility(View.GONE);

        // shows the error icon and message
        getView().findViewById(R.id.errorText).setVisibility(View.VISIBLE);
        getView().findViewById(R.id.error_image).setVisibility(View.VISIBLE);
    }


    public void listenForTransferProgress() {
        if (mProgressListener != null && !mListening && containerActivity.getFileDownloaderBinder() != null) {
            containerActivity.getFileDownloaderBinder().addDatatransferProgressListener(mProgressListener, getFile());
            mListening = true;
            setButtonsForTransferring();
        }
    }


    public void leaveTransferProgress() {
        if (mProgressListener != null && containerActivity.getFileDownloaderBinder() != null) {
            containerActivity.getFileDownloaderBinder()
                .removeDatatransferProgressListener(mProgressListener, getFile());
            mListening = false;
        }
    }


    /**
     * Helper class responsible for updating the progress bar shown for file uploading or downloading
     */
    private class ProgressListener implements OnDatatransferProgressListener {
        int mLastPercent;
        WeakReference<ProgressBar> mProgressBar;

        ProgressListener(ProgressBar progressBar) {
            mProgressBar = new WeakReference<>(progressBar);
        }

        @Override
        public void onTransferProgress(
                long progressRate, long totalTransferredSoFar, long totalToTransfer, String filename
        ) {
            int percent = (int)(100.0*((double)totalTransferredSoFar)/((double)totalToTransfer));
            if (percent != mLastPercent) {
                ProgressBar pb = mProgressBar.get();
                if (pb != null) {
                    pb.setProgress(percent);
                    pb.postInvalidate();
                }
            }
            mLastPercent = percent;
        }
    }

    public void setError(boolean error) {
        mError = error;
    }
}

File: app/src/main/java/com/owncloud/android/ui/preview/FileDownloadFragment.kt
/*
 * Nextcloud - Android Client
 *
 * SPDX-FileCopyrightText: 2026 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-FileCopyrightText: 2022 Álvaro Brey <alvaro@alvarobrey.com>
 * SPDX-FileCopyrightText: 2022 Tobias Kaminsky <tobias@kaminsky.me>
 * SPDX-FileCopyrightText: 2020 Chris Narkiewicz <hello@ezaquarii.com>
 * SPDX-FileCopyrightText: 2018 Andy Scherzinger <info@andy-scherzinger.de>
 * SPDX-FileCopyrightText: 2015 ownCloud Inc.
 * SPDX-FileCopyrightText: 2013-2015 David A. Velasco <dvelasco@solidgear.es>
 * SPDX-License-Identifier: GPL-2.0-only AND (AGPL-3.0-or-later OR GPL-2.0-only)
 */

package com.owncloud.android.ui.preview

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.nextcloud.client.account.User
import com.nextcloud.client.di.Injectable
import com.nextcloud.client.jobs.download.FileDownloadHelper.Companion.instance
import com.nextcloud.utils.extensions.getParcelableArgument
import com.nextcloud.utils.extensions.getTypedActivity
import com.owncloud.android.R
import com.owncloud.android.databinding.FileDownloadFragmentBinding
import com.owncloud.android.datamodel.OCFile
import com.owncloud.android.lib.common.utils.Log_OC
import com.owncloud.android.ui.fragment.FileFragment
import com.owncloud.android.utils.theme.ViewThemeUtils
import javax.inject.Inject

class FileDownloadFragment :
    FileFragment(),
    View.OnClickListener,
    Injectable {
    private var user: User? = null
    private var _binding: FileDownloadFragmentBinding? = null
    private val binding get() = _binding!!

    @Inject
    lateinit var viewThemeUtils: ViewThemeUtils

    private var ignoreFirstSavedState = false
    private var downloadError = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        arguments?.let {
            file = it.getParcelableArgument(ARG_FILE, OCFile::class.java)
            ignoreFirstSavedState = it.getBoolean(ARG_IGNORE_FIRST)
            user = it.getParcelableArgument(ARG_USER, User::class.java)
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        super.onCreateView(inflater, container, savedInstanceState)

        if (!ignoreFirstSavedState) {
            downloadError = requireArguments().getBoolean(EXTRA_ERROR)
            instance().downloadFile(user!!, file)
        } else {
            ignoreFirstSavedState = false
        }

        _binding = FileDownloadFragmentBinding.inflate(inflater, container, false)

        viewThemeUtils.material.run {
            colorProgressBar(binding.progressBar)
            colorMaterialTextButton(binding.cancelBtn)
            themeCardView(binding.progressCard)
            themeCardView(binding.errorCard)
        }

        binding.cancelBtn.setOnClickListener(this)
        binding.fileDownloadLL.setOnClickListener {
            getTypedActivity(PreviewImageActivity::class.java)?.toggleFullScreen()
        }

        setButtonsForTransferring(isTransferring = !downloadError)

        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.run {
            putParcelable(EXTRA_FILE, file)
            putParcelable(EXTRA_USER, user)
            putBoolean(EXTRA_ERROR, downloadError)
        }
    }

    override fun onClick(v: View) {
        if (v.id == R.id.cancelBtn) {
            containerActivity.getFileOperationsHelper().cancelTransference(file)
            requireActivity().finish()
        } else {
            Log_OC.e(TAG, "Incorrect view clicked!")
        }
    }

    private fun setButtonsForTransferring(isTransferring: Boolean) {
        val binding = _binding ?: return

        val transferringVisibility = if (isTransferring) View.VISIBLE else View.GONE
        val remoteVisibility = if (isTransferring) View.GONE else View.VISIBLE

        binding.run {
            downloadingContainer.visibility = transferringVisibility
            errorContainer.visibility = remoteVisibility

            progressText.apply {
                setText(R.string.downloader_download_in_progress_ticker)
                visibility = transferringVisibility
            }
        }
    }

    fun setError(error: Boolean) {
        downloadError = error
    }

    companion object {
        const val EXTRA_FILE: String = "FILE"
        const val EXTRA_USER: String = "USER"
        private const val EXTRA_ERROR = "ERROR"
        private const val ARG_FILE = "FILE"
        private const val ARG_IGNORE_FIRST = "IGNORE_FIRST"
        private const val ARG_USER = "USER"
        private val TAG: String = FileDownloadFragment::class.java.simpleName

        fun newInstance(file: OCFile?, user: User?, ignoreFirstSavedState: Boolean): Fragment =
            FileDownloadFragment().apply {
                arguments = Bundle().apply {
                    putParcelable(ARG_FILE, file)
                    putParcelable(ARG_USER, user)
                    putBoolean(ARG_IGNORE_FIRST, ignoreFirstSavedState)
                }
            }
    }
}
